aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,283 @@
+
+.. literalinclude:: routing/sample/RoutingSample/HelloRouter.cs
+  :language: c#
+  :emphasize-lines: 8,18
+
+``HelloRouter`` checks to see if ``RouteData`` includes a value for the key ``name``. If not, it immediately returns without handling the request. Otherwise, the request is handled (by writing out "Hi {name}!") and the ``RouteContext`` is updated to note that the request was handled. This prevents additional routes from handling the request. The ``GetVirtualPath`` method is used for :ref:`link generation <link-generation>`.
+
+.. note:: Remember, it's possible for a particular route **template** to match a given request, but the associated route **handler** can still reject it, allowing a different route to handle the request.)
+
+This route was configured to use an :ref:`inline constraint <route-constraints>`, signified by the ``:alpha`` in the name route value. This constraint limits which requests this route will handle, in this case to alphabetical values for ``name``. Thus, a request for "/hello/steve" will be handled, but a request to "/hello/123" will not (instead, the request will pass through to the "Hello World!" request delegate).
+
+Template Routes
+---------------
+The most common way to define routes is using ``TemplateRoute`` and route template strings. When a ``TemplateRoute`` matches, it calls its target ``IRouter`` handler. In a typical MVC app, you might use a default template route with a string like this one: 
+
+.. image:: /fundamentals/routing/_static/default-mvc-routetemplate.png
+
+This route template would be handled by the `MvcRouteHandler <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Infrastructure/MvcRouteHandler/index.html>`_ ``IRouter`` instance. Tokens within curly braces (``{ }``) define `route value` parameters which will be bound if the route is matched. You can define more than one route value parameter in a route segment, but they must be separated by a literal value. For example ``{controller=Home}{action=Index}`` would not be a valid route, since there is no literal value between ``{controller}`` and ``{action}``. These route value placeholders must have a name, and may have additional attributes specified.
+
+Route value parameters may have *default values*, designated by specifying the default after the parameter name, separated by an =. For example, ``controller=Home`` would define ``Home`` as the default value for ``controller``. The default value is used if no value is present in the URL for the parameter. In addition to default values, route parameters may be optional (specified by appending a ``?`` to the end of the parameter name, as in ``id?``). Route parameters may also have constraints, which further restrict which routes the template will match.
+
```

Comment:
The difference between optional and "has default" is that a route parameter with a default value always produces a value, and an optional sometimes does not :+1: 

I think the text is pretty clear already, but work this in if you feel like it adds value